include/jemalloc/internal/div.h

```diff
@@ -0,0 +1,35 @@
+#ifndef JEMALLOC_INTERNAL_DIV_H
+#define JEMALLOC_INTERNAL_DIV_H
+
+#include "jemalloc/internal/assert.h"
+
+/*
+ * This module does the division that computes the index of a region in a slab,
+ * given its offset relative to the base.
+ * That is, given a divisor d, an n = i * d (all integers), we'll return i.
+ * We do some pre-computation to do this more quickly than a CPU division
+ * instruction.
+ * We bound n < 2^32, and don't support dividing by one.
+ */
+
+typedef struct div_info_s div_info_t;
+struct div_info_s {
+	uint32_t magic;
+#ifndef NDEBUG
+	size_t d;
+#endif
+};
+
+void div_init(div_info_t *div_info, size_t divisor);
+
+static inline size_t
+div_compute(div_info_t *div_info, size_t n) {
+	assert(n <= (uint32_t)-1);
+	size_t i = ((uint64_t)n * (uint64_t)div_info->magic) >> 32;
```

Comment:
will the mult instruction be very slow on 32-bit arch in this case (If so should we consider not doing it for them)?

Reply:
In the places where I tried, the compiler is smart enough to generate a "multiply high-unsigned" instruction on 32-bit. I added a comment.